src/main/java/service/WeatherForecastService.java

```diff
@@ -1,13 +1,75 @@
 package service;
 
+import api.client.ForecastApiClient;
 import models.WeatherForecast;
-import parsers.WeatherJSONParser;
+import api.server.ForecastApiGatewayServer;
+import parsers.ForecastJSONParser;
 
 import java.io.IOException;
 
+/**
+ * Service class for retrieving and searching weather forecasts.
+ * This class provides methods to start and stop the API server, as well as searching for weather forecasts by city.
+ */
 public class WeatherForecastService {
+    private ForecastApiGatewayServer server;
+    private final ForecastApiClient client;
+    private static WeatherForecastService instance;
 
-    public static WeatherForecast searchWeatherCity(String city) throws IOException {
-        return WeatherJSONParser.parsJsonByCity(city);
+    private static WeatherForecast weatherForecast = null;
+    private static final int port = 9000;
+
+    /**
+     * Gets the singleton instance of WeatherForecastService.
+     *
+     * @throws IOException If an I/O error occurs while creating the client.
+     */
+    private WeatherForecastService() throws IOException {
+        client = new ForecastApiClient(port);
+    }
+    public static WeatherForecastService getInstance() throws IOException {
+        if (instance == null) {
+            instance = new WeatherForecastService();
+        }
+        return instance;
+    }
+
+    /**
+     * Searches for the weather forecast of a city.
+     *
+     * @param city The name of the city for which to retrieve the weather forecast.
+     * @return WeatherForecast object representing the weather information for the specified city.
+     * @throws IOException If an I/O error occurs while retrieving the weather forecast.
+     */
+    public WeatherForecast searchWeatherCity(String city) throws IOException, RuntimeException {
+            startSearchingAPI();
+            String forecastJSON = client.getJSON(city);
```

Comment:
There could be a race condition between `startSearchApi()` and `client.getJson()`. Consider the situation when the new thread started in `startSearchApi();` is suspended by the OS and client calls `client.getJSON()` before the server socket is actually opened. I think we need to have some future inside `ForecastApiGatewayServer` that is being resolved after the socket creation. The code here could look like this:
```java
    startSearchingAPI();
    waitSearchingApiIsUp();
    String forecastJSON = client.getJSON(city);

```